frontend/src/components/Table.tsx

```diff
@@ -0,0 +1,114 @@
+import { Pagination, Toolbar, ToolbarContent, ToolbarItem } from '@patternfly/react-core';
+import {
+  TableComposable,
+  Thead,
+  Tr,
+  Th,
+  TableComposableProps,
+  Caption,
+  Td,
+} from '@patternfly/react-table';
+import React, { useEffect } from 'react';
+import useTableColumnSort, { SortableData } from '~/utilities/useTableColumnSort';
+
+type TableProps<DataType> = {
+  data: DataType[];
+  columns: SortableData<DataType>[];
+  rowRenderer: (data: DataType) => React.ReactNode;
+  enablePagination?: boolean;
+  minPageSize?: number;
+  toolbarContent?: React.ReactElement<typeof ToolbarItem>;
+  emptyTableView?: React.ReactElement<typeof Tr>;
+  caption?: string;
+} & Omit<TableComposableProps, 'ref' | 'data'>;
+
+const Table = <T,>({
+  data: allData,
+  columns,
+  rowRenderer,
+  enablePagination,
+  minPageSize = 10,
+  toolbarContent,
+  emptyTableView,
+  caption,
+  ...props
+}: TableProps<T>): React.ReactElement => {
+  const [page, setPage] = React.useState(1);
+  const [pageSize, setPageSize] = React.useState(minPageSize);
+
+  const data = enablePagination ? allData.slice(pageSize * (page - 1), pageSize * page) : allData;
+
+  // update page to 1 if data changes (common when filter is applied)
+  useEffect(() => setPage(1), [data]);
+
+  const sort = useTableColumnSort<T>(columns, 0);
+
+  const showPagination = enablePagination && data.length > minPageSize;
+  const pagination = (pageDirection: 'up' | 'down') => (
+    <Pagination
+      dropDirection={pageDirection}
+      perPageComponent="button"
+      itemCount={data.length}
+      perPage={pageSize}
+      page={page}
+      onSetPage={(e, newPage) => setPage(newPage)}
+      onPerPageSelect={(e, newSize, newPage) => {
+        setPageSize(newSize);
+        setPage(newPage);
+      }}
+      widgetId="table-pagination"
+    />
+  );
+
+  return (
+    <>
+      <Toolbar>
+        <ToolbarContent>
+          {toolbarContent}
+          {showPagination && (
+            <ToolbarItem variant="pagination" alignment={{ default: 'alignRight' }}>
+              {pagination('down')}
+            </ToolbarItem>
+          )}
+        </ToolbarContent>
+      </Toolbar>
```

Comment:
Toolbar needs to not be rendered if you have no toolbarContent and you're not showing pagination.

You created header space:
![image](https://user-images.githubusercontent.com/8126518/222735983-2bfda027-db88-4eda-b108-910211762b7f.png)